relayer/chains/cosmos/message_handlers.go

```diff
@@ -0,0 +1,195 @@
+	"github.com/cosmos/relayer/v2/relayer/provider"
+	"github.com/cosmos/relayer/v2/relayer/provider/cosmos"
+	"go.uber.org/zap"
+)
+
+type MsgHandlerParams struct {
+	messageInfo   interface{}
+	foundMessages processor.ChannelMessageCache
+}
+
+var messageHandlers = map[string]func(*CosmosChainProcessor, MsgHandlerParams) bool {
+	processor.MsgTransfer:        (*CosmosChainProcessor).handleMsgTransfer,
+	processor.MsgRecvPacket:      (*CosmosChainProcessor).handleMsgRecvPacket,
+	processor.MsgAcknowledgement: (*CosmosChainProcessor).handleMsgAcknowlegement,
+	processor.MsgTimeout:         (*CosmosChainProcessor).handleMsgTimeout,
+	processor.MsgTimeoutOnClose:  (*CosmosChainProcessor).handleMsgTimeoutOnClose, 
+
+	// TODO client, connection, channel messages
+}
+
+func typedMessage[T *packetInfo | *channelInfo | *clientInfo | *connectionInfo](messageInfo interface{}) T {
+	typedInfo, ok := messageInfo.(T)
+	if !ok {
+		panic("invalid message info provided")
+	}
+	return typedInfo
+}
+
+// isPacketApplicable returns true if packet is applicable to the channels for path processors that are subscribed to this chain processor
+func (ccp *CosmosChainProcessor) isPacketApplicable(message string, packetInfo *packetInfo, foundMessages processor.ChannelMessageCache, channelKey processor.ChannelKey) bool {
+	if !ccp.isRelayedChannel(channelKey) {
+		return false
+	}
+	if _, ok := foundMessages[channelKey]; !ok {
+		return true
+	}
+	if _, ok := foundMessages[channelKey][message]; !ok {
+		return true
+	}
+	for sequence := range foundMessages[channelKey][message] {
+		if sequence == packetInfo.packet.Sequence {
+			// already have this sequence number
+			// there can be multiple MsgRecvPacket, MsgAcknowlegement, MsgTimeout, and MsgTimeoutOnClose for the same packet
+			// from different relayers.
+			return false
+		}
+	}
+
+	return true
+}
+
+// retainApplicableMessage assumes the packet is applicable to the channels for a path processor that is subscribed to this chain processor.
+// It creates cache path if it doesn't exist, then caches message.
+func retainPacketMessage(message string, packetInfo *packetInfo, foundMessages processor.ChannelMessageCache, channelKey processor.ChannelKey, ibcMsg provider.RelayerMessage) {
+	if _, ok := foundMessages[channelKey]; !ok {
+		foundMessages[channelKey] = make(processor.MessageCache)
+	}
+	if _, ok := foundMessages[channelKey][message]; !ok {
+		foundMessages[channelKey][message] = make(processor.SequenceCache)
+	}
+	foundMessages[channelKey][message][packetInfo.packet.Sequence] = ibcMsg
+}
+
+// BEGIN packet msg handlers
+func (ccp *CosmosChainProcessor) handleMsgTransfer(p MsgHandlerParams) bool {
+	packetInfo := typedMessage[*packetInfo](p.messageInfo)
+	// source chain processor will call this handler
+	// source channel used as key because MsgTransfer is sent to source chain
+	channelKey := packetInfo.channelKey()
+	if !ccp.isPacketApplicable(processor.MsgTransfer, packetInfo, p.foundMessages, channelKey) {
```

Comment:
I see `isPacketApplicable` duplicated quite a bit. 

To me, it signals importance and maybe considering making public as part of an interface for example.